src/plans/gradient_plan.jl

```diff
@@ -1,82 +1,177 @@
 
-specify a problem for gradient based algorithms.
+specify an objetive containing a cost and its gradient
 
 # Fields
-* `M`        – a manifold ``\mathcal M``
-* `cost`     – a function ``F: \mathcal M → ℝ`` to minimize
-* `gradient!!` – the gradient ``\operatorname{grad}F:\mathcal M → \mathcal T\mathcal M`` of the cost function ``F``.
 
-Depending on the [`AbstractEvaluationType`](@ref) `T` the gradient has to be provided
+* `cost`       – a function ``f\colon\mathcal M → ℝ``
+* `gradient!!` – the gradient ``\operatorname{grad}f\colon\mathcal M → \mathcal T\mathcal M``
+  of the cost function ``f``.
 
-* as a function `x -> X` that allocates memory for `X` itself for an [`AllocatingEvaluation`](@ref)
-* as a function `(X,x) -> X` that work in place of `X` for an [`MutatingEvaluation`](@ref)
+Depending on the [`AbstractEvaluationType`](@ref) `T` the gradient can have to forms
+
+* as a function `(M, p) -> X` that allocates memory for `X`, i.e. an [`AllocatingEvaluation`](@ref)
+* as a function `(M, X, p) -> X` that work in place of `X`, i.e. an [`InplaceEvaluation`](@ref)
 
 # Constructors
-    GradientProblem(M, cost, gradient; evaluation=AllocatingEvaluation())
+    ManifoldGradientObjective(cost, gradient; evaluation=AllocatingEvaluation())
 
-# See also
-[`gradient_descent`](@ref), [`GradientDescentOptions`](@ref)
+# Used with
+[`gradient_decent`](@ref), [`conjugate_gradient_descent`](@ref), [`quasi_Newton`](@ref)
 """
-struct GradientProblem{T<:AbstractEvaluationType,mT<:AbstractManifold,C,G} <:
-       AbstractGradientProblem{T}
-    M::mT
+struct ManifoldGradientObjective{T<:AbstractEvaluationType,C,G} <:
+       AbstractManifoldGradientObjective{T,C,G}
     cost::C
     gradient!!::G
 end
-function GradientProblem(
-    M::mT, cost::C, gradient::G; evaluation::AbstractEvaluationType=AllocatingEvaluation()
-) where {mT<:AbstractManifold,C,G}
-    return GradientProblem{typeof(evaluation),mT,C,G}(M, cost, gradient)
+function ManifoldGradientObjective(
+    cost::C, gradient::G; evaluation::AbstractEvaluationType=AllocatingEvaluation()
+) where {C,G}
+    return ManifoldGradientObjective{typeof(evaluation),C,G}(cost, gradient)
 end
 
+get_gradient_function(amo::AbstractManifoldGradientObjective) = amo.gradient!!
+
+@doc raw"""
+    ManifoldCostGradientObjective{T} <: AbstractManifoldObjective{T}
+
+specify an objetive containing one function to perform a combined computation of cost and its gradient
+
+# Fields
+
+* `costgrad!!` – a function that computes both the cost ``f\colon\mathcal M → ℝ``
+  and its gradient ``\operatorname{grad}f\colon\mathcal M → \mathcal T\mathcal M``
+
+Depending on the [`AbstractEvaluationType`](@ref) `T` the gradient can have to forms
+
+* as a function `(M, p) -> (c, X)` that allocates memory for the gradient `X`, i.e. an [`AllocatingEvaluation`](@ref)
+* as a function `(X, p) -> (c, X)` that work in place of `X`, i.e. an [`InplaceEvaluation`](@ref)
```

Comment:
```suggestion
* as a function `(M, X, p) -> (c, X)` that work in place of `X`, i.e. an [`InplaceEvaluation`](@ref)
```

Reply:
Ah, right, I have not yet re-read the docs. Thanks for noticing – this was a little inconsistent, since I introduced the M somewhere (from 0.2 to 0.3?) and did not check all docs at that time.